verification/spec/connection-handshake/chm.tla

```diff
@@ -0,0 +1,241 @@
+
+    Optional, add later:
+    - prefix
+    
+ ***************************************************************************)
+ConnectionEnds ==
+    [
+        connectionID : ConnectionIDs \union {Null},
+        clientID : ClientIDs \union {Null}
+    ]
+
+
+(***************************** Connections *****************************
+    A set of connection records.
+    A connection record contains the following fields:
+
+    - localEnd -- a connection end
+      Stores the connection end for the local chain.
+
+    - remoteEnd -- a connection end
+      Stores the connection end for the remote chain.
+ ***************************************************************************)
+Connections ==
+    [
+        localEnd : ConnectionEnds,
+        remoteEnd : ConnectionEnds
+    ]
+
+
+(***************************** ConnectionHandshakeModules ******************
+    A set of records defining the CHM.
+    A CHM record contains the following fields:
+
+    - connectionState -- a string
+      Stores the current state of this connection. It has one of the
+      following values: "UNINIT", "INIT", "TRYOPEN", "OPEN".
+
+    - connection -- a connection with two ends
+      Stores the connection.
+
+    - chainHeight -- a height
+      Stores the height of the chain running alongside the CHM.
+ ***************************************************************************)
+ConnectionHandshakeModules ==
+    [
+        connectionState : ConnectionStates,
+        connection : Connections,
+        chainHeight : Heights
+        \* light client, consensus state, etc.. 
+    ]
+
+
+
+(******************************** Messages ********************************
+These messages are connection handshake specific.
+ ***************************************************************************)
+ConnectionHandshakeMessages ==
+    [type : {"ConnOpenInit"}, 
+     connection : Connections]
+\*    \union
+\*    [type : {"ConnOpenTry"},
+\*        connection : Connections,
+\*        stateProof : Proofs, 
+\*        consensusHeight : Proofs]
+
+
+
+(***************************************************************************
+ Connection handshake message handlers.
+***************************************************************************)
+
+\* The chModule handles a "ConnOpenInit" message.
+connectionOpenInit(chm, initMsg) ==
+  \* If we're in the right state.
+  IF chm.connectionState = "UNINIT" THEN
+         [initCHM |-> [chm EXCEPT !.connection = initMsg.connection,
+                                  !.connectionState = "INIT"],
+          msg |-> [type |-> "ConnOpenTry",
+                   connection |-> initMsg.connection]]
+                   \* TODO: Proofs go here.                   
+   \* Otherwise, do not update the chain
+   ELSE [initCHM |-> chm,
+         msg |-> noMsg ]
+
+handleMsgConnectionOpenInit ==
+ /\ inMsg.type = "ConnOpenInit"
+ /\ LET res == connectionOpenInit(chModule, inMsg) IN
+      /\ chModule' = res.initCHM
```

Comment:
we probably need this to be a sequence (one element per height) and append to it.

Reply:
I guess this is related to your comment below "We should keep track of connection info per height." This also important also to be able to create/verify proofs, no?

Reply:
correct.